### ISSUE TITLE
Javascript heap out of memory

### ISSUE THREAD

**Issue by valotas** [pnpm/pnpm#938]
### pnpm version
1.21.0

### Code to reproduce the issue:
```
$ pnpm install
```

### Expected behavior:
install modules

### Actual behavior:
```
PS C:\dev\Projects\xxx> pnpm.cmd install
 WARN  mochify > watchify > chokidar: Skipping failed optional dependency fsevents@1.1.3
Resolving: total 818, reused 462, downloaded 114

<--- Last few GCs --->

   30372 ms: Mark-sweep 1359.3 (1436.5) -> 1357.8 (1436.5) MB, 792.2 / 0.0 ms [allocation failure] [GC in old space requested].
   31150 ms: Mark-sweep 1357.8 (1436.5) -> 1357.8 (1436.5) MB, 777.1 / 0.0 ms [allocation failure] [GC in old space requested].
   31976 ms: Mark-sweep 1357.8 (1436.5) -> 1363.6 (1412.5) MB, 825.9 / 0.0 ms [last resort gc].
   32797 ms: Mark-sweep 1363.6 (1412.5) -> 1369.4 (1412.5) MB, 821.3 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 000003D1608CFB61 <JS Object>
    1: buildTree(aka buildTree) [C:\dev\global_node_modules\pnpm-global\1\node_modules\.my.domain.com+8000\supi\0.7.5\@pnpm!logger@1.0.0\node_modules\supi\lib\api\install.js:471] [pc=00000223B3618147] (this=000003D160804381 <undefined>,ctx=00000222730E6841 <an Object with map 0000026388CC0179>,parentNodeId=00000071888F76D9 <String[628]: :/:my.domain.com+8000/fx-entry/5...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```
### Additional information:

 - `node -v` prints: `v6.9.5`
 - Windows 7

**Comment by zkochan** [comment]
Do you have this error with other package managers? (npm or Yarn)

**Comment by valotas** [comment]
npm `v5.5.1-canary.8` and `4.6.1` works just fine. I do not have yarn installed.

**Comment by zkochan** [comment]
Do you use the standard `registry.npmjs.org/` for installation?

**Comment by valotas** [comment]
No, that I can not do.

**Comment by zkochan** [comment]
Does the registry you use support corgi? It is something that `registry.npmjs.org` added this year. They return not all the meta info but only what is needed for installation.

I think pnpm currently stores in memory all the manifest/package.json info of each package. Although only some of the fields are used during installation. So probably if we won't save all the fields, RAM memory usage will be reduced.

I can check this after I implement the "aliases" feature. Probably some time next week

**Comment by valotas** [comment]
Nope, I do not think that is supported.

**Comment by valotas** [comment]
Is there anything that I can do? which are the fields needed per `package.json`?

**Comment by zkochan** [comment]
These: https://github.com/pnpm/supi/blob/master/src/types.ts#L144-L164

Maybe this line of code can be rewritten to only include the neede fields: https://github.com/pnpm/supi/blob/master/src/install/installMultiple.ts#L379

**Comment by zkochan** [comment]
Published in [v1.23.0](https://github.com/pnpm/pnpm/releases/tag/v1.23.0)

**Comment by valotas** [comment]
I am still having the problem

**Comment by valotas** [comment]
It looks like the heap is getting full in the `buildTree` recursion found in supi's `install.js`

**Comment by zkochan** [comment]
Maybe unpacking tarballs consumes a lot of memory. Try changing the value of the `network-concurrency` config https://github.com/pnpm/pnpm#network-concurrency

**Comment by valotas** [comment]
I tried 5 and got the following

```
?ERROR? Expected concurrency to be a number from 1 and up, got 5 (string)
```

To be honest though, I am not sure that a couple of missing fields could resolve the problem. Sooner or later it will strike back, if that is the problem.

It looks like the buildTree recursion is for some reson too much...

**Comment by valotas** [comment]
It looks like a getter for `children` which would lazy load it solves the problem. I can create a pr for that.

**Comment by zkochan** [comment]
the tree building is not really introducing a lot of new data. Unless the Node IDs are really long. If your solution doesn't help maybe we could somehow make the node ids smaller

**Comment by valotas** [comment]
It looks like it is enough to bring the vm down before a `Maximum call stack size exceeded` :)

**Comment by valotas** [comment]
I'll retest tomorrow with a couple of bigger (in terms of dependencies) projects.

**Comment by zkochan** [comment]
Seems like the benchmarks are a little bit worse in some cases. This change optimized memory usage but added additional cpu load. So I am not sure yet whether it is a good idea to publish it or not.

As I mentioned, the only new data in buildTree is the NodeID + a number + a boolean. Only nodeID can be big. But I think it can be substituted with a number because the Node ID should only be needed during building the tree, to detect circular dependencies. In worst case, a new number can be added to reference the parent node. But two numbers should be a lot smaller than the long Node IDs

**Comment by valotas** [comment]
How do you run the benchmarks?

**Comment by zkochan** [comment]
I use this repo: https://github.com/pnpm/node-package-manager-benchmark

I remove the last results of Yarn and run all the cases for yarn and pnpm. Then I check if differences between yarn/pnpm are the same as before.

**Comment by zkochan** [comment]
Sometimes results are random, so they have to run 3 times. The benchmarks are really time-consuming.

**Comment by valotas** [comment]
How do you run them?

```
npm install
npm run benchmark
```
?

**Comment by zkochan** [comment]
yes, but first I remove the last folder from `/results/yarn`

And I link pnpm from my local repo.

**Comment by valotas** [comment]
And how do you see memory consumption vs cpu cycles?

**Comment by zkochan** [comment]
I did not inspect it. It was my assumption because pnpm slowed down a bit, according to the results. The peers resolution part of pnpm is actually one of the slowest parts of the application and children are accessed several times there https://github.com/pnpm/supi/blob/master/src/link/resolvePeers.ts

**Comment by valotas** [comment]
Sorry, I have a hard time going through the code. As I see the generated nodeId is the key of your "map" object. I do not know how you can get rid of it, unless you do not care about, but in `resolvePeersOfNode` you are making use of it. There for I do not see how can it get substituted by a number. Or did I understand something wrong?

Apart from that, if the tree is accessed a lot, it make sense that it needs more cpu, as it is a getter and it will regenerated every time it is accessed. That is the reason I do not like getter. They do not give "feedback" to the user of what is going on behind the scenes. Let me fix that...

Finally, since you care about performance, I do not thing that objects as maps are faster than actual maps

**Comment by zkochan** [comment]
`ctx.tree` can be rewritten to an array and nodeId can be just a number that will reference the element in the array. nodeId is currently a really long piece of string, a concatenation of package ids. But it is needed only inside buildTree: https://github.com/pnpm/supi/blob/master/src/api/install.ts#L652

**Comment by valotas** [comment]
Ok, I see that is also used in https://github.com/pnpm/supi/blob/master/src/link/resolvePeers.ts#L225 but it looks like you just iterate over it, so all you need is a Set. Am I right?

**Comment by zkochan** [comment]
Yes, something that references them. Not the long node ID

The long node id is used for logging purposes here: https://github.com/pnpm/supi/blob/master/src/link/resolvePeers.ts#L284-L289

That code can be removed for now. Logging of peer warnings won't be so nice, but performance is more important

**Comment by zkochan** [comment]
@valotas I merged your fix, not published yet.

Could you create a sample project that would replicate the issue? If this is a private project, could you choose the open source dependencies from it? Maybe that will be enough. I'd use that project to test pnpm in the future

**Comment by valotas** [comment]
Unfortuantelly I can not share the repo. I will try to replicate the problem though and come back to you.

**Comment by valotas** [comment]
Never the less, thanks a lot for your help!

**Comment by zkochan** [comment]
Published with version [1.23.1](https://github.com/pnpm/pnpm/releases/tag/v1.23.1)
